crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -789,6 +799,18 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    fn infer_match_pattern_definition(
+        &mut self,
+        _pattern: &ast::Pattern,
+        _index: u32,
+        definition: Definition<'db>,
+    ) {
+        // TODO(dhruvmanila): The correct way to infer types here is to perform structural matching
+        // against the subject expression type (available in `self.index.expression_ty(subject)`)
+        // and extract the type at the `index` position if the pattern matches.
+        self.types.definitions.insert(definition, Type::Unknown);
```

Comment:
If I'm not mistaken, I think the structural matching will be dependent on the type of the subject expression.

Reply:
Yes, it will, which means we need a reference to the subject stored in the pattern Definition as well.